### DISCUSSION TITLE
Best way to convert a Unix epoch timestamp to a Skyfield time

### DISCUSSION THREAD

**Discussion by roe-dl** [skyfielders/python-skyfield#1027] [Q&A] [answered]
What is the best way to convert a Unix epoch timestamp to a Skyfield time?

By definition the Unix epoch timestamp is UTC, and the leap seconds are ignored. So the day and the time of day can simply be distinguished by dividing the value by 86400. The quotient is the day, the remainder the seconds within the day.

Assuming `ts` is the Skyfield timescale and `timestamp` is some Unix epoch timestamp, I guess, it could be:
```
ts.from_datetime(datetime.datetime.fromtimestamp(timestamp,tz=datetime.zoneinfo.utc))
```

But this seams to me complicated, as it involves using the `datetime` module and converting the timestamp twice, first to `datetime` format, and second to Skyfield `Time`.

So there is an alternative, but I am not sure, whether it is correct or not:
```
ts.utc_jd(timestamp/86400.0+2440587.5)
```

**Comment by brandon-rhodes** [comment, marked as answer]
Alas, there is no method `.utc_jd()`, because I didn't want Skyfield to have to decide how to assign floating-point JD values to a non-uniform timescale like UTC, where some days are longer than others. But instead you can use the normal calendar-oriented `.utc()` method, telling it you want a date that's hundreds or thousands of days into 1970, and since it knows that some days are longer and have leap seconds in them, it will wind up successfully creating the date you want. Thus:

```
import datetime
from skyfield.api import load, utc

ts = load.timescale()

# Example timestamp: 2025 Jan 1 00:00:00.

y = 2025 - 1970
leaps = y // 4 + 1
timestamp = 60 * 60 * 24 * (365 * y + leaps)

# The approach demonstrated in the question does work.

t = ts.from_datetime(datetime.datetime.fromtimestamp(timestamp, tz=utc))
print(t.utc_strftime())

# But instead we can submit the number of days!

t = ts.utc(1970, 1, 1 + timestamp / 86400.0)
print(t.utc_strftime())
```

The output:

```
2025-01-01 00:00:00 UTC
2025-01-01 00:00:00 UTC
```

I wonder if I should add a method that does this, to help out other folks who deal with UNIX timestamps.

**Reply by roe-dl** [reply]
When I searched the Internet before writing this question I found an article at [Stackoverflow](https://stackoverflow.com/questions/73856074/skyfield-timescale-object-does-not-include-leapseconds-during-creation-of-a-utc) which confused me. It reads:

> ```Python
> from skyfield.api import load
> load.timescale().utc(1970, 1, 1, 0, 0, 1664203669).utc_strftime()
> Out[27]: '2022-09-26 14:47:22 UTC'
> ```

But the Unix timestamp `1664203669` is `2022-09-26 14:47:49 UTC` not `2022-09-26 14:47:22 UTC`.

If I apply your suggestion to that timestamp, the correct time is displayed:

```Python
>>> from skyfield.api import load
>>> ts = load.timescale()
>>> t = ts.utc(1970,1,1+1664203669/86400.0)
>>> print(t.utc_strftime())
2022-09-26 14:47:49 UTC
```

So, thank you for clarification.
